src/storages/sled_storage/transaction.rs

```diff
@@ -76,73 +88,88 @@ impl Transaction for SledStorage {
-                return Ok(None);
+    async fn commit(self) -> MutResult<Self, ()> {
+        let (txid, created_at) = match self.state {
+            State::Transaction {
+                txid, created_at, ..
+            } => (txid, created_at),
+            State::Idle => {
+                return Err((
+                    self,
+                    Error::StorageMsg("no transaction to commit".to_owned()),
+                ));
             }
+        };
 
-            let data_items = fetch_items(key::temp_data_prefix(txid))?;
-            let schema_items = fetch_items(key::temp_schema_prefix(txid))?;
-            let index_items = fetch_items(key::temp_index_prefix(txid))?;
-
-            Ok(Some((data_items, schema_items, index_items)))
-        });
+        if let Err(e) = lock::fetch(&self.tree, txid, created_at, self.tx_timeout) {
+            return Err((self, e));
+        }
 
-        let (data_items, schema_items, index_items) = match temp_items {
-            Some(items) => items,
-            None => {
-                return Ok((
-                    Self {
-                        tree: self.tree,
-                        state: State::Idle,
-                    },
-                    (),
-                ));
+        let (storage, _) = transaction!(self, move |tree| { lock::release(tree, txid) })?;
+        let gc = || {
+            if storage.tree.get("gc_lock").map_err(err_into)?.is_some() {
+                return Ok(());
             }
-        };
 
-        transaction!(self, move |tree| {
-            for (temp_key, value_key) in data_items.iter() {
-                tree.remove(temp_key)?;
+            storage.tree.insert("gc_lock", &[1]).map_err(err_into)?;
 
-                let snapshot = tree
-                    .get(value_key)?
-                    .map(|l| bincode::deserialize(&l))
-                    .transpose()
-                    .map_err(err_into)
-                    .map_err(ConflictableTransactionError::Abort)?;
+            let gc_result = storage.gc();
 
-                let snapshot: Snapshot<Row> = match snapshot {
-                    Some(snapshot) => snapshot,
-                    None => {
-                        continue;
-                    }
-                };
+            storage.tree.remove("gc_lock").map_err(err_into)?;
 
-                match snapshot.rollback(txid) {
-                    Some(snapshot) => {
-                        let snapshot = bincode::serialize(&snapshot)
-                            .map_err(err_into)
-                            .map_err(ConflictableTransactionError::Abort)?;
+            gc_result
+        };
 
-                        tree.insert(value_key, snapshot)?;
-                    }
-                    None => {
-                        tree.remove(value_key)?;
-                    }
-                };
-            }
+        match gc() {
+            Ok(_) => Ok((storage, ())),
+            Err(e) => Err((storage, e)),
+        }
```

Comment:
Can it be implemented in this way?

```rust
gc().map(|_| (storage, ())).map_err(|e| (storage, e))
```

Reply:
This solution does not work.
Storage needs to be copied...

Reply:
```rust
gc().map(|_| (storage.clone(), ()))
            .map_err(|e| (storage, e))
```
hmm..

Reply:
Yeah, I'd also really like to make this part more simple, but closures takes ownership so it cannot work.
![image](https://user-images.githubusercontent.com/2025065/130019539-7d0f01f7-45c4-4a53-a231-fe6ef399e9df.png)

In the repo, you may recognize lots of macro uses named `try_self!` or `try_into!` and all that `try_{..}` macro is due to this issue.